### DISCUSSION TITLE
Function override

### DISCUSSION THREAD

**Discussion by fork123aniket** [pyg-team/pytorch_geometric#4440] [Q&A]
How can I override a function named `explain_message()` inside `MessagePassing` interface for each message passing layer in custom explainer which inherits from `torch.nn.Module` without touching each layer type' (GAT, GCN, RGCN, etc.) original implementation??

Currently, my `explain_message()` implements the following functionality:-

```
def explain_message((self, out: Tensor, x_i: Tensor, x_j: Tensor) -> Tensor:
      if out.size(-1) != self.process_message[0].normalized_shape[0]:
          out = self.convert(out)
      basis_messages = self.process_message(out)
      
      if self.message_scale is not None:
          basis_messages = basis_messages * self.message_scale.unsqueeze(
              -1)
      
          if self.message_replacement is not None:
              if basis_messages.shape == self.message_replacement.shape:
                  basis_messages = (basis_messages +
                                    (1 - self.message_scale).unsqueeze(
                                       -1) * self.message_replacement)
              else:
                  basis_messages = (basis_messages +
                                    ((1 - self.message_scale).unsqueeze(
                                        -1) *
                                     self.message_replacement.unsqueeze(
                                         0)))
      
      self.latest_messages = basis_messages
      self.latest_source_embeddings = x_j
      self.latest_target_embeddings = x_i
      
      return basis_messages
```

**Comment by rusty1s** [comment]
You should be able to just assign them to each layer: `conv.explain_message = my_custom_explain_message`.

**Reply by fork123aniket** [reply]
Yes, it worked!!! However, this approach of overriding `explain_message()` of `MessagePassing` interface in the `GraphMAskExplainer` class doesn't seem to help saving layer-wise `basis_messages`, `x_j`, and `x_i` values in `self.latest_messages`, `self.latest_source_embeddings`, and `self.latest_target_embeddings` respectively as here **_self_** refers to the object of `GraphMaskExplainer` class and not `MessagePassing` class of PyG, thereby every time calling `explain_message()`, defined above, always gives the same value (the last `MessagePassing` layer's `self.latest_messages`, `self.latest_source_embeddings`, and `self.latest_target_embeddings` values). Consequently, at the time of producing the explanations, I ain't able to grab `self.latest_messages`, `self.latest_source_embeddings`, and `self.latest_target_embeddings` values for each layer in the trained GNN model. Please let me know if you know any way which can help saving `basis_messages`, `x_j`, and `x_i` in their corresponding `self.latest_messages`, `self.latest_source_embeddings`, and `self.latest_target_embeddings` parameters for each layer in the model.

**Reply by rusty1s** [reply]
Can you show me an example of how you define `explain_message` and assign it?

**Reply by fork123aniket** [reply]
Below are the three functions am using in `GraphMaskExplainer` class:-

```
def __set_flags__(self, model):
    for module in model.modules():
        if isinstance(module, MessagePassing):
            module.explain = True
            self.message_scale = None
            self.message_replacement = None
            module.explain_message = self.explain_message


def explain_message(self, out, x_i, x_j):
    basis_messages = Sequential(LayerNorm(out.size(-1)), ReLU())(out)

    if self.message_scale is not None:
        basis_messages = basis_messages * self.message_scale.unsqueeze(
            -1)

        if self.message_replacement is not None:
            if basis_messages.shape == self.message_replacement.shape:
                basis_messages = (basis_messages +
                                  (1 - self.message_scale).unsqueeze(
                                      -1) * self.message_replacement)
            else:
                basis_messages = (basis_messages +
                                  ((1 - self.message_scale).unsqueeze(
                                      -1) *
                                   self.message_replacement.unsqueeze(
                                       0)))

    self.latest_messages = basis_messages
    self.latest_source_embeddings = x_j
    self.latest_target_embeddings = x_i

    return basis_messages


def __inject_messages__(self, message_scale, message_replacement, set=False):
    i = 0
    for module in self.model_to_explain.modules():
        if isinstance(module, MessagePassing):
            if not set:
                self.message_scale = message_scale[i]
                self.message_replacement = message_replacement[i]
                i = i + 1
            else:
                self.message_scale = None
                self.message_replacement = None
```

`__set_flags__()` function is used to set `explain` variable and to override `explain_message()` of `MessagePassing` with the one defined above (i.e., with `def explain_message(self, out, x_i, x_j)` function definition). With `__inject_messages__()` and `set=True` (training time), the intention is to set all the layers' `message_scale` and `message_replacement` to `None` and with `set=False` (at the time of producing explanations), `message_scale` and `message_replacement` of each of the layers of the trained GNN model must be set to `message_scale[i]` and `message_replacement[i]` respectively. Likewise, the same should be done with `self.latest_messages` `self.latest_source_embeddings`, and `self.latest_target_embeddings` defined inside custom `explain_message()` above.

The problem here is that I don't know the way to store `basis_messages`, `x_j`, and `x_i` values to `self.latest_messages` `self.latest_source_embeddings`, and `self.latest_target_embeddings` respectively. Moreover, don't know how to set `message_scale[i]` and `message_replacement[i]` to each layers' `self.message_scale` and `self.message_replacement` since here **_self_** will always let these parameters (`self.message_scale`, `self.message_replacement`, `self.latest_messages`, `self.latest_source_embeddings`, and `self.latest_target_embeddings`) to have values associated with the last layer of the GNN model and not let them to have their corresponding values for each of the layers in the GNN model.

**Reply by rusty1s** [reply]
I think you are right. It doesn't look so easy to replace this method while not losing access to `self`. Need to think about it :(
For now, feel free to make the changes directly inside `explain_message` without overriding it.

**Reply by rusty1s** [reply]
I think I solved this in https://github.com/pyg-team/pytorch_geometric/pull/4448. Please have a look :)

**Reply by fork123aniket** [reply]
@rusty1s  This change to `MessagePassing` interface doesn't seem to work either. In the below custom `explain_message()`:-

```
def explain_message(self, out, x_i, x_j):
    basis_messages = Sequential(LayerNorm(out.size(-1)), ReLU())(out)

    if self.message_scale is not None:
        basis_messages = basis_messages * self.message_scale.unsqueeze(
            -1)

        if self.message_replacement is not None:
            if basis_messages.shape == self.message_replacement.shape:
                basis_messages = (basis_messages +
                                  (1 - self.message_scale).unsqueeze(
                                      -1) * self.message_replacement)
            else:
                basis_messages = (basis_messages +
                                  ((1 - self.message_scale).unsqueeze(
                                      -1) *
                                   self.message_replacement.unsqueeze(
                                       0)))

    self.latest_messages = basis_messages
    self.latest_source_embeddings = x_j
    self.latest_target_embeddings = x_i

    return basis_messages
```

**_self_** refers to the object of `GraphMaskExplainer` and not of any layer's type (GAT, GCN, etc.), thereby throwing error: `AttributeError: 'GCNConv' object has no attribute 'latest_source_embeddings'`. I also attempted to use the test, you've written in the response to #4448 PR, but it didn't help either. Here, the problem is we need to inject **_message_** (`message_scale`) and **_learned baseline_** (`message_replacement`) in each message passing layer of the trained GNN model along with saving each of these values (`basis_messages`, `x_j`, and `x_i`) to their corresponding each message passing layer's object parameters (`latest_messages`, `latest_source_embeddings`, and `latest_target_embeddings`). I made changes to the custom `__set_flags__()` inside `GraphMaskExplainer` class, which also had gone into vein:-

```
def __set_flags__(self, model):
    for module in model.modules():
        if isinstance(module, MessagePassing):
            module.explain_message = self.explain_message.__get__(module, MessagePassing)
            module.explain = True
            self.message_scale = None
            self.message_replacement = None
```

since, here, `self.message_scale` only sets the `GraphMaskExplainer`'s `message_scale` attribute and not sets any layer type's `message_scale`, which is also true for `message_replacement`. When I superseded `self.message_scale` with `module.message_scale` inside `__set_flags__()` defined above, it threw an error `AttributeError: 'GCNConv' object has no attribute 'message_scale'`, which is also held true for `message_replacement`. So, this is not what we want.
The solution, at this point, I can think of is to make this custom `explain_message()` as `MessagePassing`'s `explain_message()` default implementation since the `GNNExplainer`'s `explain_message` functionality is quite easy and we can make it available within `gnn_explainer.py` file by overriding this function in it because `GNNExplainer` doesn't compute layer-wise explanation weights for each edge, however `GraphMaskExplainer` does. WDYT??

**Reply by rusty1s** [reply]
I am not sure I understand. `self` inside `explain_message` clearly belongs to the GNN layer if you follow the pattern in the test case. Of course, your custom explainer message should be an isolated function, not a method attached to the explainer. You also might want to check existence of attributes via `if getattr(self, message_scale, None) is not None`.

**Reply by fork123aniket** [reply]
Thanks, your recommendations made the code work properly. Now I'm able to produce explanations in a seamless way by overriding `explain_message()` of `MessagePassing` with the custom `explain_message()` defined above. To fix the problem, defined the custom `explain_message()` outside of the explainer class and used `getattr(self, 'message_scale', None)` inside this function and, **_bingo!!!!_**, it worked. Thanks so much for your quick responses and invaluable suggestions. Will soon create a **_PR_** on this!!!!

**Reply by rusty1s** [reply]
Super :) Glad that it is now working!

**Comment by 688spiderbin** [comment]
@fork123aniket
Hello, thanks for your contributions, 
I want to ask how to explain MessagePassing Model by using the function you changed or you can give me an example like the official colab notebook show how to explain nn.Module class model by using Captum.  That will be really convenient for us to use this method to do our research. 

I appreciate a lot !

**Reply by fork123aniket** [reply]
Hi @688spiderbin,

I don't know if GraphMaskExplainer has ever been made available for Captum functionality, however, you can always checkout my GitHub Repo to see an example of GraphMaskExplainer and how to explain predictions using it after April 16. Am preparing to upload the whole project as a repo, however, will be able to upload it sometime after April 16. Will notify you of this once it's available here... Thank you.
